### PR TITLE
Migrate to Girder 3

### PR DESCRIPTION
**Pull request by brianhelba**

### CODE REVIEW

**Comment by brianhelba** [comment]
@danlamanna This is rebased on top of `master` again, if you'd like to apply anything on top of it.

**Comment by danlamanna** [comment]
Starting to force push

File: ansible/roles/isic/vars/main.yml [hunk 1]
```diff
@@ -1,8 +1,7 @@
 ---
 
-girder_version: "933096b33334bb9c30c3c7a659dbab32691cba3d"
-worker_version: "042a69d6ae7d472e32242d98f8b7acb3148980f9"
-large_image_version: "4dc82976d58b2a6e6542863e2e410628649a040c"
+girder_version: "master"
```
Comment (brianhelba) [inline ansible/roles/isic/vars/main.yml]: Do either of these two variables get used anymore? Both should be indirectly installed by `pip install isic-archive`.

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        'python-dateutil',
+        'python-dotenv',
+        'requests',
+        'requests_toolbelt',
+        'scikit-image',
+        'scipy',
+        'sentry-sdk',
+        'six',
+    ],
+    extras_require={
```
Comment (brianhelba) [inline setup.py]: Let's cut this section, for reasons discussed.

File: ansible/vagrant-webservers.yml [hunk 3]
```diff
@@ -41,3 +41,15 @@
...
       isic_upload_role_arn: "fake-arn"
       localstack_data_dir: "{{ ansible_user_dir }}/localstack-data"
+
+  post_tasks:
+    - name: Install dev requirements
+      pip:
+        name: tox
+        virtualenv: "{{ python_dist_path }}"
+
+    - name: Activate virtualenv on login
```
Comment (brianhelba) [inline ansible/vagrant-webservers.yml]: @brianhelba Where did this go before?

File: ansible/vagrant-workers.yml [hunk 4]
```diff
@@ -44,74 +39,54 @@
...
       large_image_include_vips: true
 
-  post_tasks:
-    - name: Install large_image plugin
-      command: "{{ python_dist_path }}/bin/girder-install plugin --force --symlink {{ large_image_path }}"
-
-    - name: Install ISIC Archive
-      command: "{{ python_dist_path }}/bin/girder-install plugin --force --symlink {{ isic_archive_path }}"
-      notify: Reload Celery
+    - role: celery
```
Comment (brianhelba) [inline ansible/vagrant-workers.yml]: If there's a reason that the `celery` role needs to happen after `large_image`, can you add a code comment explaining it?

File: isic_archive/__init__.py [hunk 5]
```diff
@@ -74,62 +75,66 @@ def clearRouteDocs():
...
-    info['apiRoot'].study = api.StudyResource()
-    info['apiRoot'].task = api.TaskResource()
-    api.attachUserApi(info['apiRoot'].user)
+class IsicArchive(GirderPlugin):
+    DISPLAY_NAME = 'ISIC Archive'
+
+    def load(self, info):
+        sentry_sdk.init(environment=os.getenv('SENTRY_ENVIRONMENT'))
+
+        getPlugin('oauth').load(info)
```
Comment (brianhelba) [inline isic_archive/__init__.py]: @brianhelba Document why we depend on these being called first, which is so we can erase their Swagger descriptions.

File: isic_archive/__init__.py [hunk 6]
```diff
@@ -74,62 +75,66 @@ def clearRouteDocs():
...
+        getPlugin('large_image').load(info)
+
+        # set the title of the HTML pages
+        info['serverRoot'].updateHtmlVars({'title': 'ISIC Archive'})
```
Comment (brianhelba) [inline isic_archive/__init__.py]: @brianhelba Why is this still here? Shouldn't `brandName` get set via a setting?

File: isic_archive/__init__.py [hunk 7]
```diff
@@ -74,62 +75,66 @@ def clearRouteDocs():
...
+
+        # register models
+        ModelImporter.registerModel('annotation', Annotation, 'isic_archive')
```
Comment (brianhelba) [inline isic_archive/__init__.py]: @brianhelba How does this fail if absent?

File: isic_archive/__init__.py [hunk 8]
```diff
@@ -74,62 +75,66 @@ def clearRouteDocs():
...
+        ModelImporter.registerModel('segmentation', Segmentation, 'isic_archive')
+        ModelImporter.registerModel('study', Study, 'isic_archive')
+        ModelImporter.registerModel('user', User, 'isic_archive')
+
+        # add event listeners
+        events.bind('rest.get.describe/:resource.after',
+                    'onDescribeResource', onDescribeResource)
+
+        # add custom model searching
+        resource.allowedSearchTypes.update({
```
Comment (brianhelba) [inline isic_archive/__init__.py]: @brianhelba Is there a new API for this?

File: isic_archive/models/study.py [hunk 9]
```diff
@@ -35,10 +34,11 @@
 _masterFeaturesPath = os.path.normpath(os.path.join(
     os.path.dirname(__file__), '..', '..', 'isic-archive-gui', 'src', 'masterFeatures.json'
 ))
-with open(_masterFeaturesPath, 'rb') as _masterFeaturesStream:
-    MASTER_FEATURES = [
-        feature['id'] for feature in json.load(_masterFeaturesStream)
-    ]
+# with open(_masterFeaturesPath, 'rb') as _masterFeaturesStream:
```
Comment (brianhelba) [inline isic_archive/models/study.py]: TODO

File: isic_archive/tasks/__init__.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+from dotenv import load_dotenv
+
+load_dotenv()
```
Comment (brianhelba) [inline isic_archive/tasks/__init__.py]: Consider promoting this to the top level of `isic_archive` (not just `tasks`). Also consider making an `isic_archive.settings` (or similar) where this lives.

File: isic_archive/tasks/app.py [hunk 11]
```diff
@@ -49,26 +52,25 @@ class CeleryAppConfig(object):
     # jsonpickle is used to support passing object ids between tasks
     task_serializer = 'jsonpickle'
 
+    # TODO: Remove when Celery version >= 4.3 since it then infers from env vars
+    result_backend = os.getenv('CELERY_RESULT_BACKEND')
```
Comment (brianhelba) [inline isic_archive/tasks/app.py]: This could potentially go away when GW is upgraded / dropped.

File: isic_archive/tasks/app.py [hunk 12]
```diff
@@ -49,26 +52,25 @@ class CeleryAppConfig(object):
...
-def fixGirderImports(sender, **kwargs):
-    from girder.utility.server import configureServer
-    configureServer(curConfig={
-        'server': {
-            'mode': 'production'
-        },
-        'cherrypy_server': False
-    })
+def addMailTemplates(sender, **kwargs):
+    mail_utils.addTemplateDirectory(
```
Comment (brianhelba) [inline isic_archive/tasks/app.py]: Can you add a comment of something like "This is a bare-bones version of `IsicArchive.load`".

File: plugin.cmake [hunk 13]
```diff
@@ -1,63 +0,0 @@
...
-  )
-  set_property(TEST "eslint_${name}" PROPERTY LABELS girder_browser)
-endfunction()
-
-add_standard_plugin_tests(NO_SERVER)
-
-# Fetch external data
-set(_data_files "")
-foreach(_data_file
-  "plugins/isic_archive/test_1_small_1.jpg"
```
Comment (brianhelba) [inline plugin.cmake]: Remove these files.

File: plugin.json [hunk 14]
```diff
@@ -1,7 +0,0 @@
-{
-    "name": "ISIC Archive",
-    "description": "The ISIC dermoscopy image archive.",
-    "url": "https://github.com/ImageMarkup/isic-archive",
-    "version": "1.0.0",
-    "dependencies": ["gravatar", "large_image", "oauth", "worker"]
```
Comment (brianhelba) [inline plugin.json]: We may want to drop `gravatar` in the future, but let's not make this PR change it. So, we should add `girder_gravatar` to this plugin's dependencies.

File: setup.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        'dev': [
+            'tox'
+        ]
+    },
+    dependency_links=[
```
Comment (brianhelba) [inline setup.py]: No need for this. We can track alpha releases to PyPI.

File: tox.ini [hunk 16]
```diff
@@ -1,6 +1,16 @@
...
+[testenv]
+# TODO: Remove networkx dep when girder-worker stops depending on it
+# TODO: Remove celery dep when https://github.com/girder/girder_worker/issues/330 is resolved
+# TODO: Figure out why per-process tests are necessary in user_test
+deps = pytest
+       git+https://github.com/girder/girder.git#egg=pytest-girder&subdirectory=pytest_girder
+       celery==4.1.1
+       networkx==1.9.0
+       pytest-xdist
+commands = pytest --boxed --verbose --showlocals {posargs} plugin_tests
```
Comment (brianhelba) [inline tox.ini]: We could rename `plugin_tests` to `test`.

File: .circleci/config.yml [hunk 17]
```diff
@@ -89,87 +9,47 @@ jobs:
...
+      - checkout:
+          path: /home/circleci/project
 
       - run:
           name: Run Tox
           command: tox
-          working_directory: girder/plugins/isic_archive
-
-      - run:
-          name: Upload coverage
```
Comment (brianhelba) [inline .circleci/config.yml]: Let's see if we can unpin Coverage from `pytest_girder` and re-enable coverage downstream here.

**Comment by danlamanna** [comment]
@brianhelba this is ready for review. The only thing left is to convert the discourse SSO plugin.

**Comment by brianhelba** [comment]
:shipit:

**Pull request merged**